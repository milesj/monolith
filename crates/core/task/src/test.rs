use crate::{FileGroup, Target, Task};
use moon_config::TaskConfig;
use moon_utils::string_vec;
use rustc_hash::FxHashMap;

pub fn create_file_groups_config() -> FxHashMap<String, Vec<String>> {
    let mut map = FxHashMap::default();

    map.insert(
        String::from("static"),
        string_vec![
            "file.ts",
            "dir",
            "dir/other.tsx",
            "dir/subdir",
            "dir/subdir/another.ts",
        ],
    );

    map.insert(String::from("dirs_glob"), string_vec!["**/*"]);

    map.insert(String::from("files_glob"), string_vec!["**/*.{ts,tsx}"]);

    map.insert(String::from("globs"), string_vec!["**/*.{ts,tsx}", "*.js"]);

    map.insert(String::from("no_globs"), string_vec!["config.js"]);

    map
}

pub fn create_file_groups() -> FxHashMap<String, FileGroup> {
    let mut map = FxHashMap::default();

    map.insert(
        String::from("static"),
        FileGroup::new(
            "static",
            string_vec![
                "file.ts",
                "dir",
                "dir/other.tsx",
                "dir/subdir",
                "dir/subdir/another.ts",
            ],
        ),
    );

    map.insert(
        String::from("dirs_glob"),
        FileGroup::new("dirs_glob", string_vec!["**/*"]),
    );

    map.insert(
        String::from("files_glob"),
        FileGroup::new("files_glob", string_vec!["**/*.{ts,tsx}"]),
    );

    map.insert(
        String::from("globs"),
        FileGroup::new("globs", string_vec!["**/*.{ts,tsx}", "*.js"]),
    );

    map.insert(
        String::from("no_globs"),
        FileGroup::new("no_globs", string_vec!["config.js"]),
    );

    map
}

pub fn create_initial_task(config: Option<TaskConfig>) -> Task {
    Task::from_config(
        Target::new("project", "task").unwrap(),
        &config.unwrap_or_default(),
    )
    .unwrap()
}
