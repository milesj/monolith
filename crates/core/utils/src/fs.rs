use async_recursion::async_recursion;
use futures::future::try_join_all;
use json_comments::StripComments;
use moon_error::{map_io_to_fs_error, map_json_to_error, map_yaml_to_error, MoonError};
use regex::Regex;
use serde::de::DeserializeOwned;
use serde::Serialize;
use std::io::Read;
use std::path::{Path, PathBuf};
use std::time::{Duration, SystemTime};
use tokio::fs;

#[inline]
pub fn clean_json<T: AsRef<str>>(json: T) -> Result<String, MoonError> {
    let json = json.as_ref();

    // Remove comments
    let mut stripped = String::with_capacity(json.len());

    StripComments::new(json.as_bytes())
        .read_to_string(&mut stripped)
        .map_err(MoonError::Unknown)?;

    // Remove trailing commas
    let stripped = Regex::new(r",(?P<valid>\s*})")
        .unwrap()
        .replace_all(&stripped, "$valid");

    Ok(String::from(stripped))
}

#[inline]
pub async fn copy_file<S: AsRef<Path>, D: AsRef<Path>>(from: S, to: D) -> Result<(), MoonError> {
    let from = from.as_ref();
    let to = to.as_ref();
    let to_dir = to.parent().unwrap();

    create_dir_all(to_dir).await?;

    fs::copy(from, to)
        .await
        .map_err(|e| map_io_to_fs_error(e, from.to_path_buf()))?;

    Ok(())
}

#[async_recursion]
pub async fn copy_dir_all<T: AsRef<Path> + Send>(
    from_root: T,
    from: T,
    to_root: T,
) -> Result<(), MoonError> {
    let from_root = from_root.as_ref();
    let from = from.as_ref();
    let to_root = to_root.as_ref();
    let entries = read_dir(from).await?;
    let mut files = vec![];
    let mut dirs = vec![];

    for entry in entries {
        let path = entry.path();

        if path.is_file() {
            files.push(copy_file(
                path.to_owned(),
                to_root.join(path.strip_prefix(from_root).unwrap()),
            ));
        } else {
            dirs.push(path);
        }
    }

    // Copy files before dirs incase an error occurs
    try_join_all(files).await?;

    // Copy dirs in sequence for the same reason
    for dir in dirs {
        copy_dir_all(from_root, &dir, to_root).await?;
    }

    Ok(())
}

#[inline]
pub async fn create_dir_all<T: AsRef<Path>>(path: T) -> Result<(), MoonError> {
    let path = path.as_ref();

    if !path.exists() {
        fs::create_dir_all(&path)
            .await
            .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;
    }

    Ok(())
}

#[inline]
pub fn find_upwards<F, P>(name: F, dir: P) -> Option<PathBuf>
where
    F: AsRef<str>,
    P: AsRef<Path>,
{
    let dir = dir.as_ref();
    let findable = dir.join(name.as_ref());

    if findable.exists() {
        return Some(findable);
    }

    match dir.parent() {
        Some(parent_dir) => find_upwards(name, parent_dir),
        None => None,
    }
}

#[inline]
pub async fn metadata<T: AsRef<Path>>(path: T) -> Result<std::fs::Metadata, MoonError> {
    let path = path.as_ref();

    fs::metadata(path)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))
}

#[inline]
pub async fn read_dir<T: AsRef<Path>>(path: T) -> Result<Vec<fs::DirEntry>, MoonError> {
    let path = path.as_ref();
    let handle_error = |e| map_io_to_fs_error(e, path.to_path_buf());

    let mut entries = fs::read_dir(path).await.map_err(handle_error)?;
    let mut results = vec![];

    while let Some(entry) = entries.next_entry().await.map_err(handle_error)? {
        results.push(entry);
    }

    Ok(results)
}

#[async_recursion]
pub async fn read_dir_all<T: AsRef<Path> + Send>(path: T) -> Result<Vec<fs::DirEntry>, MoonError> {
    let path = path.as_ref();
    let handle_error = |e| map_io_to_fs_error(e, path.to_path_buf());

    let mut entries = fs::read_dir(path).await.map_err(handle_error)?;
    let mut results = vec![];

    while let Some(entry) = entries.next_entry().await.map_err(handle_error)? {
        if let Ok(file_type) = entry.file_type().await {
            if file_type.is_dir() {
                results.extend(read_dir_all(&entry.path()).await?);
            } else {
                results.push(entry);
            }
        }
    }

    Ok(results)
}

#[inline]
pub async fn read<T: AsRef<Path>>(path: T) -> Result<String, MoonError> {
    let path = path.as_ref();
    let data = fs::read_to_string(path)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

    Ok(data)
}

#[inline]
pub async fn read_json<P, D>(path: P) -> Result<D, MoonError>
where
    P: AsRef<Path>,
    D: DeserializeOwned,
{
    let path = path.as_ref();
    let contents = read_json_string(path).await?;

    let json: D =
        serde_json::from_str(&contents).map_err(|e| map_json_to_error(e, path.to_path_buf()))?;

    Ok(json)
}

#[inline]
pub async fn read_json_string<T: AsRef<Path>>(path: T) -> Result<String, MoonError> {
    clean_json(read(path).await?)
}

#[inline]
pub async fn read_yaml<P, D>(path: P) -> Result<D, MoonError>
where
    P: AsRef<Path>,
    D: DeserializeOwned,
{
    let path = path.as_ref();
    let contents = fs::read_to_string(path)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

    let json: D =
        serde_yaml::from_str(&contents).map_err(|e| map_yaml_to_error(e, path.to_path_buf()))?;

    Ok(json)
}

#[inline]
pub async fn remove<T: AsRef<Path>>(path: T) -> Result<(), MoonError> {
    let path = path.as_ref();

    if path.is_file() {
        remove_file(path).await?;
    } else if path.is_dir() {
        remove_dir_all(path).await?;
    }

    Ok(())
}

#[inline]
pub async fn remove_file<T: AsRef<Path>>(path: T) -> Result<(), MoonError> {
    let path = path.as_ref();

    if path.exists() {
        fs::remove_file(&path)
            .await
            .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;
    }

    Ok(())
}

#[inline]
pub async fn remove_dir_all<T: AsRef<Path>>(path: T) -> Result<(), MoonError> {
    let path = path.as_ref();

    if path.exists() {
        fs::remove_dir_all(&path)
            .await
            .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;
    }

    Ok(())
}

pub type RemoveDirContentsResult = (usize, u64);

pub async fn remove_dir_stale_contents<P: AsRef<Path>>(
    dir: P,
    duration: Duration,
) -> Result<RemoveDirContentsResult, MoonError> {
    let mut files_deleted: usize = 0;
    let mut bytes_saved: u64 = 0;
    let threshold = SystemTime::now() - duration;

    for entry in read_dir(dir.as_ref()).await? {
        let path = entry.path();

        if path.is_file() {
            let mut bytes = 0;

            if let Ok(metadata) = entry.metadata().await {
                bytes = metadata.len();

                if let Ok(filetime) = metadata.accessed().or_else(|_| metadata.created()) {
                    if filetime > threshold {
                        // Not stale yet
                        continue;
                    }
                } else {
                    // Not supported in environment
                    continue;
                }
            }

            if remove_file(path).await.is_ok() {
                files_deleted += 1;
                bytes_saved += bytes;
            }
        }
    }

    Ok((files_deleted, bytes_saved))
}

#[inline]
pub async fn rename<F: AsRef<Path>, T: AsRef<Path>>(from: F, to: T) -> Result<(), MoonError> {
    let from = from.as_ref();

    fs::rename(from, to.as_ref())
        .await
        .map_err(|e| map_io_to_fs_error(e, from.to_path_buf()))?;

    Ok(())
}

#[inline]
pub async fn write<T: AsRef<Path>>(path: T, data: impl AsRef<[u8]>) -> Result<(), MoonError> {
    let path = path.as_ref();

    fs::write(path, data)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

    Ok(())
}

#[inline]
pub async fn write_json<P, D>(path: P, json: &D, pretty: bool) -> Result<(), MoonError>
where
    P: AsRef<Path>,
    D: ?Sized + Serialize,
{
    let path = path.as_ref();
    let data = if pretty {
        serde_json::to_string_pretty(&json).map_err(|e| map_json_to_error(e, path.to_path_buf()))?
    } else {
        serde_json::to_string(&json).map_err(|e| map_json_to_error(e, path.to_path_buf()))?
    };

    fs::write(path, data)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

    Ok(())
}

#[inline]
pub async fn write_yaml<P, D>(path: P, yaml: &D) -> Result<(), MoonError>
where
    P: AsRef<Path>,
    D: ?Sized + Serialize,
{
    let path = path.as_ref();
    let data =
        serde_yaml::to_string(&yaml).map_err(|e| map_yaml_to_error(e, path.to_path_buf()))?;

    fs::write(path, data)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

    Ok(())
}

pub mod temp {
    use super::*;
    use moon_constants::CONFIG_DIRNAME;
    use std::fs; // TEMPORARILY sync

    pub fn get_dir() -> PathBuf {
        crate::get_workspace_root()
            .join(CONFIG_DIRNAME)
            .join("cache/temp")
    }

    pub fn get_file(source: &str, ext: &str) -> PathBuf {
        get_dir().join(format!("{:x}.{}", md5::compute(source), ext))
    }

    pub fn read<P: AsRef<Path>>(path: P) -> Result<Option<String>, MoonError> {
        let file = path.as_ref();

        if !file.exists() {
            return Ok(None);
        }

        // Temp files only last for 4 hours (half a workday)
        let threshold = SystemTime::now() - Duration::from_secs(60 * 60 * 4);

        if let Ok(metadata) = file.metadata() {
            if let Ok(filetime) = metadata.created() {
                if filetime > threshold {
                    fs::remove_file(file)?;

                    return Ok(None);
                }
            }
        }

        Ok(Some(fs::read_to_string(file)?))
    }

    pub fn write<P: AsRef<Path>, D: AsRef<str>>(path: P, data: D) -> Result<(), MoonError> {
        let path = path.as_ref();

        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)?;
        }

        fs::write(path, data.as_ref())?;

        Ok(())
    }
}

pub mod sync {
    use super::*;
    use std::fs::read_to_string;

    #[inline]
    pub fn read<P>(path: P) -> Result<String, MoonError>
    where
        P: AsRef<Path>,
    {
        let path = path.as_ref();
        let data = read_to_string(path).map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

        Ok(data)
    }

    #[inline]
    pub fn read_json<P, D>(path: P) -> Result<D, MoonError>
    where
        P: AsRef<Path>,
        D: DeserializeOwned,
    {
        let path = path.as_ref();
        let contents = read_json_string(path)?;

        let json: D = serde_json::from_str(&contents)
            .map_err(|e| map_json_to_error(e, path.to_path_buf()))?;

        Ok(json)
    }

    #[inline]
    pub fn read_json_string<T: AsRef<Path>>(path: T) -> Result<String, MoonError> {
        clean_json(read(path.as_ref())?)
    }

    #[inline]
    pub fn read_yaml<P, D>(path: P) -> Result<D, MoonError>
    where
        P: AsRef<Path>,
        D: DeserializeOwned,
    {
        let path = path.as_ref();
        let contents =
            read_to_string(path).map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

        let json: D = serde_yaml::from_str(&contents)
            .map_err(|e| map_yaml_to_error(e, path.to_path_buf()))?;

        Ok(json)
    }
}
