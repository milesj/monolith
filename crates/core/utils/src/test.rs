use crate::path;
use crate::process::output_to_string;
use std::env;
use std::path::{Path, PathBuf};
use std::process::Command;

pub use assert_fs::TempDir;

pub fn run_git_command<P, F>(dir: P, handler: F)
where
    P: AsRef<Path>,
    F: FnOnce(&mut Command),
{
    let mut cmd = Command::new(if cfg!(windows) { "git.exe" } else { "git" });
    cmd.current_dir(dir.as_ref());

    handler(&mut cmd);

    let out = cmd.output().unwrap_or_else(|e| {
        panic!("{:#?}", e);
    });

    if !out.status.success() {
        println!("{}", output_to_string(&out.stdout));
        eprintln!("{}", output_to_string(&out.stderr));
    }
}

pub fn create_sandbox<T: AsRef<str>>(fixture: T) -> assert_fs::TempDir {
    use assert_fs::prelude::*;

    let temp_dir = assert_fs::TempDir::new().unwrap();

    temp_dir
        .copy_from(get_fixtures_dir(fixture), &["**/*"])
        .unwrap();

    temp_dir
        .copy_from(get_fixtures_root(), &["shared-workspace.yml"])
        .unwrap();

    temp_dir
}

pub fn create_sandbox_with_git<T: AsRef<str>>(fixture: T) -> assert_fs::TempDir {
    use assert_fs::prelude::*;

    let temp_dir = create_sandbox(fixture);

    if !temp_dir.path().join(".gitignore").exists() {
        temp_dir
            .child(".gitignore")
            .write_str("node_modules")
            .unwrap();
    }

    // Initialize a git repo so that VCS commands work
    run_git_command(temp_dir.path(), |cmd| {
        cmd.args(["init", "--initial-branch", "master"]);
    });

    // We must also add the files to the index
    run_git_command(temp_dir.path(), |cmd| {
        cmd.args(["add", "--all", "."]);
    });

    // And commit them... this seems like a lot of overhead?
    run_git_command(temp_dir.path(), |cmd| {
        cmd.args(["commit", "-m", "Fixtures"])
            .env("GIT_AUTHOR_NAME", "moon tests")
            .env("GIT_AUTHOR_EMAIL", "fakeemail@moonrepo.dev")
            .env("GIT_COMMITTER_NAME", "moon tests")
            .env("GIT_COMMITTER_EMAIL", "fakeemail@moonrepo.dev");
    });

    temp_dir
}

pub fn get_fixtures_root() -> PathBuf {
    let mut root = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    root.push("../../../tests/fixtures");

    path::normalize(root)
}

pub fn get_fixtures_dir<T: AsRef<str>>(dir: T) -> PathBuf {
    get_fixtures_root().join(dir.as_ref())
}

pub fn replace_fixtures_dir<T: AsRef<str>, P: AsRef<Path>>(value: T, dir: P) -> String {
    let dir_str = dir.as_ref().to_str().unwrap();

    // Replace both forward and backward slashes
    value
        .as_ref()
        .replace(dir_str, "<WORKSPACE>")
        .replace(&path::standardize_separators(dir_str), "<WORKSPACE>")
}

pub fn create_moon_command<T: AsRef<Path>>(path: T) -> assert_cmd::Command {
    let mut cmd = assert_cmd::Command::cargo_bin("moon").unwrap();
    cmd.current_dir(path);
    cmd.timeout(std::time::Duration::from_secs(90));
    // Let our code know were running tests
    cmd.env("MOON_TEST", "true");
    // Hide install output as it disrupts testing snapshots
    cmd.env("MOON_TEST_HIDE_INSTALL_OUTPUT", "true");
    // Standardize file system paths for testing snapshots
    cmd.env("MOON_TEST_STANDARDIZE_PATHS", "true");
    // Enable logging for code coverage
    cmd.env("MOON_LOG", "trace");
    cmd
}

pub fn get_assert_output(assert: &assert_cmd::assert::Assert) -> String {
    get_assert_stdout_output(assert) + &get_assert_stderr_output_clean(assert)
}

pub fn get_assert_stderr_output_clean(assert: &assert_cmd::assert::Assert) -> String {
    let mut output = String::new();

    // We need to always show logs for proper code coverage,
    // but this breaks snapshots, and as such, we need to manually
    // filter out log lines and env vars!
    for line in get_assert_stderr_output(assert).split('\n') {
        if !line.starts_with("[error")
            && !line.starts_with("[ warn")
            && !line.starts_with("[ info")
            && !line.starts_with("[debug")
            && !line.starts_with("[trace")
            && !line.starts_with("  MOON_")
            && !line.starts_with("  NODE_")
        {
            output.push_str(line);
            output.push('\n');
        }
    }

    output
}

pub fn get_assert_stderr_output(assert: &assert_cmd::assert::Assert) -> String {
    String::from_utf8(assert.get_output().stderr.to_owned()).unwrap()
}

pub fn get_assert_stdout_output(assert: &assert_cmd::assert::Assert) -> String {
    String::from_utf8(assert.get_output().stdout.to_owned()).unwrap()
}

pub fn debug_sandbox_files(dir: &Path) {
    for entry in std::fs::read_dir(dir).unwrap() {
        let path = entry.unwrap().path();

        if path.is_dir() {
            debug_sandbox_files(&path);
        } else {
            println!("- {}", path.to_string_lossy());
        }
    }
}

pub fn debug_sandbox(fixture: &assert_fs::TempDir, assert: &assert_cmd::assert::Assert) {
    // List all files in the sandbox
    println!("sandbox:");
    debug_sandbox_files(fixture.path());
    println!("\n");

    // Debug outputs
    println!("stdout:\n{}\n", get_assert_stdout_output(assert));
    println!("stderr:\n{}\n", get_assert_stderr_output(assert));
    println!("status: {:#?}", assert.get_output().status);
}
