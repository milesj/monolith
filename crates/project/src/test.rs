use crate::errors::ProjectError;
use crate::file_group::FileGroup;
use crate::target::Target;
use crate::task::Task;
use crate::token::{TokenResolver, TokenSharedData};
use moon_config::TaskConfig;
use std::collections::HashMap;
use std::path::Path;

pub fn create_file_groups_config() -> HashMap<String, Vec<String>> {
    let mut map = HashMap::new();

    map.insert(
        String::from("static"),
        vec![
            "file.ts".to_owned(),
            "dir".to_owned(),
            "dir/other.tsx".to_owned(),
            "dir/subdir".to_owned(),
            "dir/subdir/another.ts".to_owned(),
        ],
    );

    map.insert(String::from("dirs_glob"), vec!["**/*".to_owned()]);

    map.insert(String::from("files_glob"), vec!["**/*.{ts,tsx}".to_owned()]);

    map.insert(
        String::from("globs"),
        vec!["**/*.{ts,tsx}".to_owned(), "*.js".to_owned()],
    );

    map.insert(String::from("no_globs"), vec!["config.js".to_owned()]);

    map
}

pub fn create_file_groups(project_root: &Path) -> HashMap<String, FileGroup> {
    let mut map = HashMap::new();

    map.insert(
        String::from("static"),
        FileGroup::new(
            "static",
            vec![
                "file.ts".to_owned(),
                "dir".to_owned(),
                "dir/other.tsx".to_owned(),
                "dir/subdir".to_owned(),
                "dir/subdir/another.ts".to_owned(),
            ],
            project_root,
        ),
    );

    map.insert(
        String::from("dirs_glob"),
        FileGroup::new("dirs_glob", vec!["**/*".to_owned()], project_root),
    );

    map.insert(
        String::from("files_glob"),
        FileGroup::new("files_glob", vec!["**/*.{ts,tsx}".to_owned()], project_root),
    );

    map.insert(
        String::from("globs"),
        FileGroup::new(
            "globs",
            vec!["**/*.{ts,tsx}".to_owned(), "*.js".to_owned()],
            project_root,
        ),
    );

    map.insert(
        String::from("no_globs"),
        FileGroup::new("no_globs", vec!["config.js".to_owned()], project_root),
    );

    map
}

pub fn create_expanded_task(
    workspace_root: &Path,
    project_root: &Path,
    config: Option<TaskConfig>,
) -> Result<Task, ProjectError> {
    let mut task = Task::from_config(
        Target::format("project", "task").unwrap(),
        &config.unwrap_or_default(),
    );
    let file_groups = create_file_groups(project_root);
    let metadata = TokenSharedData::new(&file_groups, workspace_root, project_root);

    task.expand_inputs(TokenResolver::for_inputs(&metadata))?;
    task.expand_outputs(TokenResolver::for_outputs(&metadata))?;
    task.expand_args(TokenResolver::for_args(&metadata))?; // Must be last

    Ok(task)
}
