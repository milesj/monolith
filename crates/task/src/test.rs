use crate::{FileGroup, Target, Task, TaskError, TokenResolver, TokenSharedData};
use moon_config::{ProjectConfig, TaskConfig};
use moon_utils::string_vec;
use std::collections::HashMap;
use std::path::Path;

pub fn create_file_groups_config() -> HashMap<String, Vec<String>> {
    let mut map = HashMap::new();

    map.insert(
        String::from("static"),
        string_vec![
            "file.ts",
            "dir",
            "dir/other.tsx",
            "dir/subdir",
            "dir/subdir/another.ts",
        ],
    );

    map.insert(String::from("dirs_glob"), string_vec!["**/*"]);

    map.insert(String::from("files_glob"), string_vec!["**/*.{ts,tsx}"]);

    map.insert(String::from("globs"), string_vec!["**/*.{ts,tsx}", "*.js"]);

    map.insert(String::from("no_globs"), string_vec!["config.js"]);

    map
}

pub fn create_file_groups() -> HashMap<String, FileGroup> {
    let mut map = HashMap::new();

    map.insert(
        String::from("static"),
        FileGroup::new(
            "static",
            string_vec![
                "file.ts",
                "dir",
                "dir/other.tsx",
                "dir/subdir",
                "dir/subdir/another.ts",
            ],
        ),
    );

    map.insert(
        String::from("dirs_glob"),
        FileGroup::new("dirs_glob", string_vec!["**/*"]),
    );

    map.insert(
        String::from("files_glob"),
        FileGroup::new("files_glob", string_vec!["**/*.{ts,tsx}"]),
    );

    map.insert(
        String::from("globs"),
        FileGroup::new("globs", string_vec!["**/*.{ts,tsx}", "*.js"]),
    );

    map.insert(
        String::from("no_globs"),
        FileGroup::new("no_globs", string_vec!["config.js"]),
    );

    map
}

pub fn create_initial_task(config: Option<TaskConfig>) -> Task {
    Task::from_config(
        Target::format("project", "task").unwrap(),
        &config.unwrap_or_default(),
    )
}

pub fn create_expanded_task(
    workspace_root: &Path,
    project_root: &Path,
    config: Option<TaskConfig>,
) -> Result<Task, TaskError> {
    let mut task = create_initial_task(config);
    let file_groups = create_file_groups();
    let project_config = ProjectConfig::new(project_root);
    let metadata =
        TokenSharedData::new(&file_groups, workspace_root, project_root, &project_config);

    task.expand_env(&metadata)?;
    task.expand_inputs(TokenResolver::for_inputs(&metadata))?;
    task.expand_outputs(TokenResolver::for_outputs(&metadata))?;
    task.expand_args(TokenResolver::for_args(&metadata))?; // Must be last

    Ok(task)
}
