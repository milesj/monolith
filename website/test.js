const readline = require('readline');

console.log('START');

const rl = readline.createInterface({
	input: process.stdin,
	output: process.stdout,
});

function ask(question) {
	console.log('ASK');
	rl.question(question, (answer) => {
		console.log('Q');
		rl.write(`The answer received:  ${answer}\n`);

		ask(question);
	});
}

ask('What is your name: ');
