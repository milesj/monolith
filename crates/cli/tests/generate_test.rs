use insta::assert_snapshot;
use moon_utils::test::{create_moon_command, create_sandbox, get_assert_output};
use predicates::prelude::*;
use std::fs;

fn get_path_safe_output(assert: &assert_cmd::assert::Assert) -> String {
    get_assert_output(assert).replace('\\', "/")
}

#[test]
fn creates_a_new_template() {
    let fixture = create_sandbox("generator");

    let assert = create_moon_command(fixture.path())
        .arg("generate")
        .arg("new-name")
        .arg("--template")
        .assert();
    let output = get_path_safe_output(&assert);

    assert!(predicate::str::contains("Created a new template new-name at").eval(&output));
    assert!(fixture.path().join("templates/new-name").exists());

    assert.success();
}

#[test]
fn generates_files_from_template() {
    let fixture = create_sandbox("generator");

    let assert = create_moon_command(fixture.path())
        .arg("generate")
        .arg("standard")
        .arg("./test")
        .assert();

    assert_snapshot!(get_path_safe_output(&assert));

    assert!(fixture.path().join("test").exists());
    assert!(fixture.path().join("test/file.ts").exists());
    assert!(fixture.path().join("test/folder/nested-file.ts").exists());
    assert!(!fixture.path().join("test/template.yml").exists());
}

#[test]
fn doesnt_generate_files_when_dryrun() {
    let fixture = create_sandbox("generator");

    let assert = create_moon_command(fixture.path())
        .arg("generate")
        .arg("standard")
        .arg("./test")
        .arg("--dryRun")
        .assert();

    assert_snapshot!(get_path_safe_output(&assert));

    assert!(!fixture.path().join("test").exists());
    assert!(!fixture.path().join("test/file.ts").exists());
    assert!(!fixture.path().join("test/folder/nested-file.ts").exists());
    assert!(!fixture.path().join("test/template.yml").exists());
}

#[test]
fn overwrites_existing_files_when_forced() {
    let fixture = create_sandbox("generator");

    create_moon_command(fixture.path())
        .arg("generate")
        .arg("standard")
        .arg("./test")
        .assert();

    let assert = create_moon_command(fixture.path())
        .arg("generate")
        .arg("standard")
        .arg("./test")
        .arg("--force")
        .assert();

    assert_snapshot!(get_path_safe_output(&assert));

    assert!(fixture.path().join("test").exists());
    assert!(fixture.path().join("test/file.ts").exists());
    assert!(fixture.path().join("test/folder/nested-file.ts").exists());
    assert!(!fixture.path().join("test/template.yml").exists());
}

#[test]
fn overwrites_existing_files_when_interpolated_path() {
    let fixture = create_sandbox("generator");

    create_moon_command(fixture.path())
        .arg("generate")
        .arg("vars")
        .arg("./test")
        .arg("--defaults")
        .assert();

    let assert = create_moon_command(fixture.path())
        .arg("generate")
        .arg("vars")
        .arg("./test")
        .arg("--defaults")
        .arg("--force")
        .assert();

    assert_snapshot!(get_path_safe_output(&assert));

    // file-[stringNotEmpty]-[number].txt
    assert!(fixture.path().join("./test/file-default-0.txt").exists());
}

#[test]
fn renders_and_interpolates_templates() {
    let fixture = create_sandbox("generator");

    let assert = create_moon_command(fixture.path())
        .arg("generate")
        .arg("vars")
        .arg("./test")
        .arg("--defaults")
        .assert();

    assert.success();

    assert_snapshot!(fs::read_to_string(fixture.path().join("./test/expressions.txt")).unwrap());
    assert_snapshot!(fs::read_to_string(fixture.path().join("./test/control.txt")).unwrap());
}

#[test]
fn renders_with_custom_vars_via_args() {
    let fixture = create_sandbox("generator");

    let assert = create_moon_command(fixture.path())
        .arg("generate")
        .arg("vars")
        .arg("./test")
        .arg("--defaults")
        .arg("--")
        .args([
            "--no-boolTrue",
            "--boolFalse",
            "--string=abc",
            "--stringNotEmpty",
            "xyz",
            "--number=123",
            "--numberNotEmpty",
            "456",
            "--enum=c",
            "--multenumNotEmpty",
            "a",
        ])
        .assert();

    assert.success();

    assert_snapshot!(fs::read_to_string(fixture.path().join("./test/expressions.txt")).unwrap());
    assert_snapshot!(fs::read_to_string(fixture.path().join("./test/control.txt")).unwrap());
}

#[test]
fn interpolates_destination_path() {
    let fixture = create_sandbox("generator");

    let assert = create_moon_command(fixture.path())
        .arg("generate")
        .arg("vars")
        .arg("./test")
        .arg("--defaults")
        .assert();

    // Verify output paths are correct
    assert_snapshot!(get_path_safe_output(&assert));

    // file-[stringNotEmpty]-[number].txt
    assert!(fixture.path().join("./test/file-default-0.txt").exists());
}

#[test]
fn errors_when_parsing_custom_var_types() {
    let fixture = create_sandbox("generator");

    let assert = create_moon_command(fixture.path())
        .arg("generate")
        .arg("vars")
        .arg("./test")
        .arg("--defaults")
        .arg("--")
        .arg("--number=abc")
        .assert();

    assert_snapshot!(get_path_safe_output(&assert));
}

#[test]
fn supports_custom_filters() {
    let fixture = create_sandbox("generator");

    let assert = create_moon_command(fixture.path())
        .arg("generate")
        .arg("vars")
        .arg("./test")
        .arg("--defaults")
        .assert();

    assert.success();

    assert_snapshot!(fs::read_to_string(fixture.path().join("./test/filters.txt")).unwrap());
}

mod frontmatter {
    use super::*;

    #[test]
    fn changes_dest_path() {
        let fixture = create_sandbox("generator");

        let assert = create_moon_command(fixture.path())
            .arg("generate")
            .arg("frontmatter")
            .arg("./test")
            .arg("--defaults")
            .assert();

        assert.success();

        assert!(!fixture.path().join("./test/to.txt").exists());
        assert!(fixture.path().join("./test/to-NEW.txt").exists());
        assert_snapshot!(fs::read_to_string(fixture.path().join("./test/to-NEW.txt")).unwrap());
    }

    #[test]
    fn force_writes_file() {
        let fixture = create_sandbox("generator");

        fs::create_dir_all(fixture.path().join("test")).unwrap();
        fs::write(fixture.path().join("test/forced.txt"), "Original content").unwrap();

        let assert = create_moon_command(fixture.path())
            .arg("generate")
            .arg("frontmatter")
            .arg("./test")
            .arg("--defaults")
            .assert();

        assert.success();

        assert_snapshot!(fs::read_to_string(fixture.path().join("./test/forced.txt")).unwrap());
    }

    #[test]
    fn skips_over_file() {
        let fixture = create_sandbox("generator");

        let assert = create_moon_command(fixture.path())
            .arg("generate")
            .arg("frontmatter")
            .arg("./test")
            .arg("--defaults")
            .assert();

        assert.success();

        assert!(!fixture.path().join("./test/skipped.txt").exists());
    }

    #[test]
    fn supports_component_vars() {
        let fixture = create_sandbox("generator");

        let assert = create_moon_command(fixture.path())
            .arg("generate")
            .arg("frontmatter")
            .arg("./test")
            .arg("--defaults")
            .assert();

        assert.success();

        assert!(fixture
            .path()
            .join("./test/components/SmallButton.tsx")
            .exists());
        assert_snapshot!(fs::read_to_string(
            fixture.path().join("./test/components/SmallButton.tsx")
        )
        .unwrap());
    }
}
