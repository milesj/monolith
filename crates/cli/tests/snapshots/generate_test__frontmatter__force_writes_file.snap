---
source: crates/cli/tests/generate_test.rs
assertion_line: 249
expression: "fs::read_to_string(fixture.path().join(\"./test/forced.txt\")).unwrap()"
---
Always write!

